lib/ringbuf/src/lib.rs

```diff
@@ -280,36 +307,115 @@ pub struct Ringbuf<T: Copy + PartialEq, const N: usize> {
     pub buffer: [RingbufEntry<T>; N],
 }
 
-impl<T: Copy + PartialEq, const N: usize> Ringbuf<T, { N }> {
-    pub fn entry(&mut self, line: u16, payload: T) {
-        let ndx = match self.last {
-            None => 0,
-            Some(last) => {
-                let ent = &mut self.buffer[last];
+///
+/// A ring buffer of parametrized type and size, plus counters tracking the
+/// total number of times each entry variant has been recorded.
+///
+/// Event counts are incremented each time an entry is added to this ring
+/// buffer. Counts are still tracked when the `disabled` feature is enabled. In
+/// order to be counted, the entry type must implement the [`Count`] trait.
+///
+/// In practice, instantiating this directly is strange -- see the
+/// [`counted_ringbuf!`] macro.
+///
+pub struct CountedRingbuf<T: Count + Copy + PartialEq, const N: usize> {
+    /// A ring buffer of the `N` most recent entries recorded by this
+    /// `CountedRingbuf`.
+    #[cfg(not(feature = "disabled"))]
+    pub ringbuf: StaticCell<Ringbuf<T, N>>,
+
+    /// Counts of the total number of times each variant of `T` has been
+    /// recorded, as defined by `T`'s [`Count`] impl.
+    pub counters: T::Counters,
+}
+
+///
+/// A countable ringbuf event.
+///
+/// This trait can (and generally should) be derived for an `enum`
+/// type using the [`#[derive(Count)]`][drv] attribute.
+///
+/// [drv]: ringbuf_macros::Count
+pub trait Count {
+    type Counters;
 
-                if ent.line == line && ent.payload == payload {
-                    // Only reuse this entry if we don't overflow the
-                    // count.
-                    if let Some(new_count) = ent.count.checked_add(1) {
-                        ent.count = new_count;
-                        return;
-                    }
-                }
+    const NEW_COUNTERS: Self::Counters;
 
-                if last + 1 >= self.buffer.len() {
-                    0
-                } else {
-                    last + 1
+    /// Increment the counter for this event.
+    fn count(&self, counters: &Self::Counters);
+}
+
+pub trait RecordEntry<T: Copy + PartialEq> {
+    fn record_entry(&self, line: u16, payload: T);
+}
```

Comment:
This is turned into a trait in order to just have one `ringbuf_entry!` macro that can record entries in a counted ringbuf _or_ a ringbuf without counts.